### PR TITLE
Use `ts-bridge` as build tool, set `moduleResolution` to `NodeNext`

### LINKED ISSUE #21: Use `ts-bridge` as build tool, set `moduleResolution` to `NodeNext`
## Explanation

As part of our OKR for upgrading TypeScript to v5.0+ (Q2 2024 O3KR4), we are updating our core dependencies to generate builds and type declarations for both CJS and ESM.

## References

- Same update made to `@metamask/utils`: https://github.com/MetaMask/utils/pull/182
- Required to update core repo's `moduleResolution` setting to `NodeNext`:
  - Blocks: https://github.com/MetaMask/core/issues/3651

### PR DESCRIPTION
**Pull request by MajorLift**
## Explanation

As part of our OKR for upgrading TypeScript to v5.0+ (Q2 2024 O3KR4), we are updating our core dependencies to generate builds and type declarations for both CJS and ESM.

## References

- Closes: https://github.com/MetaMask/superstruct/issues/21
- Same update made to `@metamask/utils`: https://github.com/MetaMask/utils/pull/182
- Required to update core repo's `moduleResolution` setting to `NodeNext`:
  - Blocks: https://github.com/MetaMask/core/issues/3651
 
### Added

- Newly exports the following types: `AnyStruct`, `EnumSchema`, `InferStructTuple`, `IsExactMatch`, `IsMatch`, `IsRecord`, `IsTuple`, `ObjectSchema`, `OmitBy`, `Optionalize`, `PickBy`, `Simplify`, `UnionToIntersection`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,24 +43,29 @@
   },
   "license": "MIT",
   "sideEffects": false,
+  "type": "module",
```
Comment (mcmire) [inline package.json]: Do we need to make this an ESM package? I think that using `ts-bridge` and upgrading the `moduleResolution` should work without this, no?

File: .eslintrc.cjs [hunk 2]
```diff
@@ -7,6 +7,9 @@ module.exports = {
     {
       files: ['*.ts'],
       extends: ['@metamask/eslint-config-typescript'],
+      rules: {
+        'import/no-useless-path-segments': ['error', { noUselessIndex: false }],
```
Comment (mcmire) [inline .eslintrc.cjs]: What is this for?

File: package.json (hunk 1 above)
Comment (MajorLift) [reply package.json]: Reverted here: https://github.com/MetaMask/superstruct/pull/18/commits/6c4e6cdae9a38476d758036c664067e88bfe9232

File: .eslintrc.cjs (hunk 2 above)
Comment (MajorLift) [reply .eslintrc.cjs]: Reverted here: https://github.com/MetaMask/superstruct/pull/18/commits/220fe62cae5da33108b2be348d87d53e05365168

**Comment by MajorLift** [comment]
@mcmire I removed the `"type": "module"` field in package.json for now. There was an ESM-only dependency (`lodash-es`) that was causing an error so I just replaced it with `lodash`.

**Review by mcmire — APPROVED** [review]
Looks good. Checked out this branch and `yarn build` locally and can confirm that this creates a different set of files than was there before.

**Pull request merged**